cou/steps/plan.py

```diff
@@ -52,24 +54,162 @@
+        )
+
+
+def is_highest_release_achieved(analysis_result: Analysis) -> None:
+    """Check if the highest OpenStack release is reached by the ubuntu series.
+
+    :param analysis_result: Analysis result.
+    :type analysis_result: Analysis
+    :raises HighestReleaseAchieved: When the OpenStack release is the last supported by the series.
+    """
+    if (
+        (current_os_release := analysis_result.current_cloud_os_release)
+        and (current_series := analysis_result.current_cloud_series)
+        and str(current_os_release) == LTS_TO_OS_RELEASE[current_series][-1]
+    ):
+        raise HighestReleaseAchieved(
+            f"No upgrades available for OpenStack {str(current_os_release).capitalize()} on "
+            f"Ubuntu {current_series.capitalize()}.\nNewer OpenStack releases "
+            "may be available after upgrading to a later Ubuntu series."
+        )
+
+
+def is_data_plane_ready_to_upgrade(analysis_result: Analysis) -> None:
+    """Check if data plane is ready to upgrade.
+
+    To be able to upgrade data-plane, first all control plane apps should be upgraded.
+
+    :param analysis_result: Analysis result
+    :type analysis_result: Analysis
+    :raises DataPlaneCannotUpgrade: When data-plane is not ready to upgrade.
+    """
+    if not analysis_result.min_os_version_data_plane:
+        raise DataPlaneCannotUpgrade(
+            "Cannot find data-plane apps. Is this a valid OpenStack cloud?"
+        )
+    if not is_control_plane_upgraded(analysis_result):
+        raise DataPlaneCannotUpgrade("Please, upgrade control-plane before data-plane")
+
+
+def is_control_plane_upgraded(analysis_result: Analysis) -> bool:
+    """Check if control plane is already upgraded.
+
+    Control-plane will be considered as upgraded when the OpenStack version of it
+    is bigger than the data-plane.
+
+    :param analysis_result: Analysis result
+    :type analysis_result: Analysis
+    :return: Whether the control plane is already upgraded or not.
+    :rtype: bool
+    """
+    control_plane = analysis_result.min_os_version_control_plane
+    data_plane = analysis_result.min_os_version_data_plane
+
+    return bool(control_plane and data_plane and control_plane > data_plane)
+
+
+def determine_upgrade_target(analysis_result: Analysis) -> OpenStackRelease:
+    """Determine the target release to upgrade to.
+
+    :param analysis_result: Analysis result.
+    :type analysis_result: Analysis
+    :raises NoTargetError: When cannot find target to upgrade
+    :raises OutOfSupportRange: When the upgrade scope is not supported
+        by the current series.
+    :return: The target OS release to upgrade the cloud to.
+    :rtype: OpenStackRelease
+    """
+    if (
+        (current_os_release := analysis_result.current_cloud_os_release)
+        and (current_series := analysis_result.current_cloud_series)
+        and not (target := current_os_release.next_release)
+    ):
```

Comment:
Since this is used also outside of this if + for return we should define it before it. 
```suggestion
    target = current_os_release.next_release
    if (
        (current_os_release := analysis_result.current_cloud_os_release)
        and (current_series := analysis_result.current_cloud_series)
        and not target
    ):
```

Reply:
type checker will complain that `current_os_release` can be None and won't have `next_release`. It's a question of choosing add a `ignore` because is already checked on other function or leave this way that is more bullet proof.

Reply:
I do not understand how this is more "bullet proof"? I think it is the same approach, but yours is less readable.
The whole idea behind assignment expression is to use it with function e.g. `if target := foo():` so the output is catch. However you are using it for properties and I do not see any benefit of it, it's only less redable.

Reply:
Your solution don't pass mypy. if we define `target` before we need to check that `current_os_release` is not None.

If the walrus operator makes less readable we can define `current_os_release`, `current_series` above and check target.

Reply:
I know, but mypy is just not able to check in assignment expression, you are not checking that is not none. So finding a way around mypy is not "better" solution. If you really want to fix that issue mypy is pointing out, you should check that it is not None also here.

Reply:
I don't see how the change @rgildein suggested would work, calling `current_os_release` before the variable is defined.

For the mypy error, would passing in `current_os_release` as parameter of the function with type check solves the problem? That is:
 ```python
def determine_upgrade_target(current_os_release: OpenStackRelease, current_series: str):
```

Reply:
>  but mypy is just not able to check in assignment expression,  you are not checking that is not none

Where did you see this? It looks like they support since [2019](https://github.com/python/mypy/pull/6899). It's also on the changelog from their [blog](https://mypy-lang.blogspot.com/2019/09/mypy-730-released.html)

I don't see the problem. checking a variable like `if my_var:` is enough to be sure that it's not None and continue.

@agileshaw I don't see the benefit of having a function like this. probably we will use like this:
 ```python
 determine_upgrade_target(analysis_result.current_cloud_os_release, analysis_result.current_cloud_series)
 ```
 So the asignament of the function needs to be:
 ```python
 def determine_upgrade_target(current_os_release: Optional[OpenStackRelease], current_series: Optional[str]):
 ```

Reply:
I see. To me this is not a blocker, but if you choose to use `# type: ignore` at the end of the function, it would be better to add a comment explaining we have already checked the type before calling this function.

Reply:
@agileshaw You are right me suggestion will not work. 
I saw that it should support it, but tell me what is the difference between 
```python
    if (
        (current_os_release := analysis_result.current_cloud_os_release)
        and (current_series := analysis_result.current_cloud_series)
        and not (target := current_os_release.next_release)
    ):
```
and
```python
    current_os_release = analysis_result.current_cloud_os_release
    current_series = analysis_result.current_cloud_series
    target = current_os_release.next_release
    
    if not target:
        ...
```
Did you check that parameter is None somewhere? IMO if mypy is not raising any error, is due that it could not check properly types of those porperties.

Reply:
I missed that walrus operator is doing also checking, so it means:
```python
if current_os_release and current_series and not target:
    ...
```
What is fine I guess, since these variables will not be empty string, list, etc., or do they?

Reply:
@agileshaw Samuel did a good comment about changing the name of the functions and I  ended up taking your approach. This way we don't need to use the `# type: ignore` at the end of the function and if the current structure we need to check twice, maybe it's not a bad deal checking once in the `determine_upgrade_target`. We will need to add another checks, so the structure is ready for checking azs, machines, hostnames and etc.